### ISSUE TITLE
piping output --color=always not respected

### ISSUE THREAD

**Issue by gabbagabbahey** [rmarquis/pacaur#213]
For example, the line below will produce colored output from Repos and uncolored output from AUR.
```
pacaur -Ss --color=always nvidia | grep --color=never "nvidia"
```

**Comment by thx1111** [comment]
Hmm - ok, reading at "man bash", under SHELL BUILTIN COMMANDS, at "getopts":
"optstring contains the option characters to be recognized; if a character is followed by a colon, the option is expected to have an argument, which should be  separated  from it by white space."

Well, that is "clear as mud", since I notice that, where pacaur has "-:", the command does *not* seem to require a space, "- ", and is accepting "--option".

Over at "http://wiki.bash-hackers.org/howto/getopts_tutorial", we read:
"Note that getopts is not able to parse GNU-style long options (--myoption) or XF86-style long options (-myoption)!"

And then:
"this option has an associated option argument (an argument to the option -f): /etc/mybackup.conf. The option argument is usually the argument following the option itself, but that isn't mandatory. Joining the option and option argument into a single argument -f/etc/mybackup.conf is valid."

It seems that the man page saying "should be  separated" is just an annoying use of the word "should".  So then, in pacaur, there is "--option", which is the "undocumented" first "-", then the option character "-", then the "missing" optional white space, then the option argument "option".

So, when there is the option "--color=always", there is the undocumented "-", the option character "-", the missing optional white space, and the option argument "color=always", which is to say, *literally* the string "color=always".  pacaur would have to parse for this string explicitly.  We then try this, *in addition to parsing for "color" alone*, but cannot use any following "${!OPTIND}", so just write-out "always" explicitly:

color=always) color=always; pacopts+=("--color=always") ; coweropts+=("--color=always"); shift;;

BTW, noting that "pacman" is quite accepting of both "--color=always" and "--color always", even though this is not documented.

That seems to work.  Give it a try and see what you think.  And you might want to also add lines to parse explicitly for the literal strings "color=never" and "color=auto", and for anything else that might take an "--option=something" form.  cower has many of these, and also is intolerant of the form using "space" instead of "=".

Hmm - pacaur could change its policy and document a requirement that any "--option" argument always requires an intervening "space" character before the argument, so that the bash script could always use the "${!OPTIND}" parameter.  That is sort of implied in the man page, but not stated.  In particular, I notice that you did *not* tell us that using the "--option=argument" form is not allowed in pacaur!  That might make things more predictable, interfacing with cower.  I don't know if people have actually run into trouble with any of the other cower option arguments.  I suppose they are not much used.

Looking at the cower man page, and comparing to what pacaur is parsing, I see:  "--by=what", "--color=when", "--domain=fqdn", "--ignore=pkg", "--sort=key", and "--rsort=key".  pacaur would have to require a "space" after the "=" anyway, when the argument is not from a small known set of arguments that can be searched exhaustively.

Then again, you could "sneek in" the "color=always)" case, alone, to avoid a certain number of annoying bug reports.  Ha!

**Comment by rmarquis** [comment]
A better solution seems to be documented [here](http://stackoverflow.com/questions/402377/using-getopts-in-bash-shell-script-to-get-long-and-short-command-line-options/6946864#6946864) (see "portable and native POSIX shell").

**Comment by thx1111** [comment]
There is this caution:
"In this example, -b (and its long form, --bravo) has a mandatory option (note the manual reconstruction of enforcing that for the long form)."

It looks like there are then three distinct cases to be addressed, not including errors:
 "--option=something"
 "--option something"
 "--option --next-option"

The "=" then still requires a distinct and separate pattern clause in the case command, no matter which way it is handled.  But then, yes, it looks like "${OPTARG#*=}" is your friend.

**Comment by AladW** [comment]
Not sure why you'd not simply go with util-linux getopt, which handles spaces correctly; I don't see people running pacaur on BSD or Solaris just yet.

**Comment by rmarquis** [comment]
Fixed in a1a28dd530e9cbefeb034a4ac32f0743a357eaaf using the trick described in the link above.

Note: this "portable and native POSIX shell" mention was to refer to the specific paragraph of the link, but portability to other OS has obviously never been an issue here.

Also, I tought about using getopt a long time ago, but eventually went with getopts due to some difficulties. I honestly don't remember the exact issue, but getopts does the job.
